### PR TITLE
osd: dissect and abstract RMWPipeline from ECBackend for sharing it with crimson

### PR DESCRIPTION
**Pull request by rzarzynski**
## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/osd/ECBackend.cc [hunk 1]
```diff
@@ -29,9 +29,9 @@
 
 #define dout_context cct
 #define dout_subsys ceph_subsys_osd
-#define DOUT_PREFIX_ARGS this
-#undef dout_prefix
-#define dout_prefix _prefix(_dout, this)
+//#define DOUT_PREFIX_ARGS this
+//#undef dout_prefix
+//#define dout_prefix _prefix(_dout, this)
```
Comment (athanatos) [inline src/osd/ECBackend.cc:34]: Can we just remove these?

File: src/osd/ECBackend.cc [hunk 2]
```diff
@@ -2220,24 +2163,10 @@ bool ECBackend::RMWPipeline::try_reads_to_commit()
...
-    // important: get_parent()->send_message_osd_cluster()
-    &waiting_reads=this->waiting_reads,
-    &waiting_commit=this->waiting_commit,
-    &pipeline_state=this->pipeline_state,
-    &cache=this->cache,
-    &completed_to=this->completed_to,
-    &committed_to=this->committed_to,
-    &tid_to_op_map=this->tid_to_op_map,
-    cct=(CephContext*)nullptr
-  ] {
```
Comment (athanatos) [inline src/osd/ECBackend.cc]: Can we just squash this into the prior commit?  It's not particularly useful for review.

File: src/osd/ECBackend.cc [hunk 3]
```diff
@@ -1161,30 +1161,30 @@ void ECBackend::handle_sub_write_reply(
   const ECSubWriteReply &op,
   const ZTracer::Trace &trace)
 {
-  map<ceph_tid_t, Op>::iterator i = rmw_pipeline.tid_to_op_map.find(op.tid);
+  map<ceph_tid_t, std::unique_ptr<Op>>::iterator i = rmw_pipeline.tid_to_op_map.find(op.tid);
```
Comment (athanatos) [inline src/osd/ECBackend.cc]: Op::ref or Ref<Op>?

File: src/osd/ECBackend.cc [hunk 4]
```diff
@@ -1557,7 +1593,10 @@ void ECBackend::submit_transaction(
   )
 {
   ceph_assert(!rmw_pipeline.tid_to_op_map.count(tid));
-  Op *op = &(rmw_pipeline.tid_to_op_map[tid]);
+  auto concete_op = std::make_unique<ECClassicalOp>();
```
Comment (athanatos) [inline src/osd/ECBackend.cc]: concrete

File: src/osd/ECBackend.cc [hunk 5]
```diff
@@ -1541,6 +1541,42 @@ void ECBackend::dump_recovery_info(Formatter *f) const
...
+      pg_t pgid,
+      const ECUtil::stripe_info_t &sinfo,
+      const std::map<hobject_t,extent_map> &partial_extents,
+      std::vector<pg_log_entry_t> &entries,
+      std::map<hobject_t,extent_map> *written,
+      std::map<shard_id_t, ObjectStore::Transaction> *transactions,
+      std::set<hobject_t> *temp_added,
+      std::set<hobject_t> *temp_removed,
+      DoutPrefixProvider *dpp,
+      const ceph_release_t require_osd_release) override
```
Comment (athanatos) [inline src/osd/ECBackend.cc:1559]: final?

File: src/osd/ECBackend.cc [hunk 6]
```diff
@@ -1541,6 +1541,42 @@ void ECBackend::dump_recovery_info(Formatter *f) const
...
+  {
+    if (t) {
```
Comment (athanatos) [inline src/osd/ECBackend.cc:1561]: Can t actually be null?

File: src/osd/ECBackend.cc [hunk 7]
```diff
@@ -1572,8 +1611,23 @@ void ECBackend::submit_transaction(
   if (client_op) {
     op->trace = client_op->pg_trace;
   }
+  op->plan = ECTransaction::get_write_plan(
```
Comment (athanatos) [inline src/osd/ECBackend.cc:1611]: For symmetry, seems like get_write_plan should also be a an ECBackend::Op method.

File: src/osd/ECTransaction.h [hunk 8]
```diff
@@ -178,11 +177,11 @@ namespace ECTransaction {
 	       (!plan.to_read.at(i.first).empty() &&
 		!i.second.has_source()));
       });
-    plan.t = std::move(t);
     return plan;
   }
 
   void generate_transactions(
+    PGTransaction* _t,
```
Comment (athanatos) [inline src/osd/ECTransaction.h:184]: We assert non-null, probably better to pass as a reference?

File: src/osd/ECBackend.cc [hunk 9]
```diff
@@ -2041,21 +2078,19 @@ bool ECBackend::RMWPipeline::try_reads_to_commit()
...
-      op->log_entries,
-      &written,
-      &trans,
-      &(op->temp_added),
-      &(op->temp_cleared),
-      get_parent()->get_dpp(),
-      get_osdmap()->require_osd_release);
-  }
+  op->generate_transactions(
+    op->plan,
```
Comment (athanatos) [inline src/osd/ECBackend.cc:2081]: We can probably remove the members from the param list?

**Review by athanatos — CHANGES_REQUESTED** [review]
Seems reasonable, let's use a Op::Ref or Ref<Op> rather than explicitly using std::unique_ptr<Op> each time, though.

File: src/osd/ECBackend.cc (hunk 2 above)
Comment (rzarzynski) [reply src/osd/ECBackend.cc]: Sure, will squash.

File: src/osd/ECBackend.cc (hunk 7 above)
Comment (rzarzynski) [reply src/osd/ECBackend.cc:1611]: Yes.

File: src/osd/ECBackend.cc (hunk 9 above)
Comment (rzarzynski) [reply src/osd/ECBackend.cc]: Fixed.

File: src/osd/ECTransaction.h (hunk 8 above)
Comment (rzarzynski) [reply src/osd/ECTransaction.h:184]: Added a new commit for that.

File: src/osd/ECBackend.cc (hunk 7 above)
Comment (rzarzynski) [reply src/osd/ECBackend.cc]: Added a `static` wrapper.

File: src/osd/ECBackend.cc (hunk 6 above)
Comment (rzarzynski) [reply src/osd/ECBackend.cc]: Nope, it's already dereferencec without extra checks.

File: src/osd/ECBackend.cc (hunk 5 above)
Comment (rzarzynski) [reply src/osd/ECBackend.cc]: Fixed.

File: src/osd/ECBackend.cc (hunk 4 above)
Comment (rzarzynski) [reply src/osd/ECBackend.cc]: Fixed.

File: src/osd/ECBackend.cc (hunk 2 above)
Comment (rzarzynski) [reply src/osd/ECBackend.cc]: Squashed.

**Comment by rzarzynski** [comment]
@athanatos: would you mind taking another look?

File: src/osd/ECBackend.cc [hunk 10]
```diff
@@ -1536,6 +1544,50 @@ void ECBackend::dump_recovery_info(Formatter *f) const
...
+  static ECTransaction::WritePlan get_write_plan(
+    const ECUtil::stripe_info_t &sinfo,
+    PGTransaction& t,
+    F &&get_hinfo,
+    DoutPrefixProvider *dpp)
+  {
+    return ECTransaction::get_write_plan(
+      sinfo,
+      t,
+      std::move(get_hinfo),
```
Comment (rzarzynski) [inline src/osd/ECBackend.cc:1586]: `std::forward`.

Comment (rzarzynski) [reply src/osd/ECBackend.cc]: Fixed.

**Review by athanatos — APPROVED** [review]

File: src/osd/ECBackend.cc [hunk 11]
```diff
@@ -1556,22 +1556,21 @@ struct ECClassicalOp : ECBackend::RMWPipeline::Op {
       DoutPrefixProvider *dpp,
       const ceph_release_t require_osd_release) final
   {
-    if (t) {
```
Comment (rzarzynski) [inline src/osd/ECBackend.cc]: Oops, I think I discovered where the t-empty ops are born:

```cpp
bool ECBackend::RMWPipeline::try_finish_rmw()
{
  // ...
    if (get_osdmap()->require_osd_release >= ceph_release_t::kraken) {
      if (op->version > get_parent()->get_log().get_can_rollback_to() &&
          waiting_reads.empty() &&
          waiting_commit.empty()) {
        // submit a dummy transaction to kick the rollforward
        auto tid = get_parent()->get_tid();
        Op *nop = &(tid_to_op_map[tid]);
        nop->hoid = op->hoid;
        nop->trim_to = op->trim_to;
        nop->roll_forward_to = op->version;
        nop->tid = tid;
        nop->reqid = op->reqid;
        waiting_reads.push_back(*nop);
      }
 ```

Comment (athanatos) [reply src/osd/ECBackend.cc]: Ah, I see.  Ok, add a comment to that effect so that future readers won't be confused.

**Comment by ljflores** [comment]
Rados approved: https://tracker.ceph.com/projects/rados/wiki/MAIN#httpstrellocomcJxeRJYse1822-wip-yuri4-testing-2023-08-10-1739

**Pull request merged**